### PR TITLE
Update self-signed cert code to use SubjectAltName

### PR DESCRIPTION
**Pull request by thezoggy**
Addresses https://github.com/sabnzbd/sabnzbd/issues/798

### CODE REVIEW

File: sabnzbd/utils/certgen.py [hunk 1]
```diff
@@ -12,33 +12,24 @@
 from cryptography.x509.oid import NameOID
 import datetime
 import os
-import struct
+from sabnzbd.getipaddress import localipv4, ipv6
```
Comment (Safihre) [inline sabnzbd/utils/certgen.py]: So clearly it's not a standalone module anymore, maybe we can just move it out of the `utils` folder and into the `sabnzbd` folder?
And then also move the `create_https_certificates` function into the file.

File: sabnzbd/utils/certgen.py [hunk 2]
```diff
@@ -53,20 +44,33 @@ def generate_key(key_size=2048, output_file='key.pem'):
...
             encryption_algorithm=serialization.NoEncryption()
+            # encryption_algorithm=serialization.BestAvailableEncryption(b"passphrase")
         ))
 
     return private_key
 
 
-def generate_local_cert(private_key, days_valid=356, output_file='cert.cert', LN='', ON='', CN=''):
+# Ported from cryptography docs/x509/tutorial.rst
+def generate_local_cert(private_key, days_valid=3560, output_file='cert.cert', LN=u'SABnzbd', ON=u'SABnzbd', CN=u'localhost'):
```
Comment (Safihre) [inline sabnzbd/utils/certgen.py]: Maybe not pre-fill SAB defaults to keep it a tiny bit general? :)

Comment (Safihre) [inline sabnzbd/utils/certgen.py]: Since the `CN` is not used, it can be removed as a parameter probably.

**Comment by Safihre** [comment]
Could you also test the default situations,  so with host set to `127.0.0.1` or `0.0.0.0`: if the apps are happier with the new certificates generated like that  than the old style ones?

File: sabnzbd/utils/certgen.py (hunk 2 above)
Comment (thezoggy) [inline sabnzbd/utils/certgen.py]: certgen is ran for one of 3 scenarios..  we do on refresh.. we do on startup.. user can manually from the dir.  putting the defaults here ensures that all 3 cases result in the same cert.

File: sabnzbd/utils/certgen.py (hunk 1 above)
Comment (thezoggy) [inline sabnzbd/utils/certgen.py]: the main scenario is sab kicking off the cert gen process..  tried to keep the code as close to the tutorial as possible (which is what we had before).. and the changes we made to the file locally I just cleaned up so it covers the scenario if the user wanted to run it manually and less likely to get different results

**Comment by thezoggy** [comment]
i can test other variations of the host being set, but the main thing is how does the user access sab.. do they go to `https://localhost` or `https://127.0.0.1` or what..  localhost is automatically added.. so that scenario is covered.. then whatever sab has as their local ip.. we add that.. so thats covered as well.

**Comment by thezoggy** [comment]
if i tell sab host is `0.0.0.0`, it then relaunch web browser to `https://127.0.0.1:8092` then telling it to generate new cert.. the SAN list does not change (not that I expected it to), since the computers local ip didnt change.. I end up with the same entries, localhost and my local ip.

So now that sab is bounded to `0.0.0.0` instead of ` ` does that change how the cert is validated against? no. because apps would still only validate off the SAN list...

Also to note, radarr does not like sab source reporting back `develop`, reported this: https://github.com/Radarr/Radarr/issues/769

I can see that if I put in `127.0.0.1` for the access ip it fails to validate but if i use the local ip or localhost.. its fine... so in the end the certs stuff is the same as before just handles the case when people access sab via their local ip or as localhost.. which covers the majority of use cases. For those that are trying to do self-signed certs for remote.. they need to generate their own/bind to a proper dns name/use reverse proxy/etc.. all outside the scope of our self signed certs.

**Comment by Safihre** [comment]
Thanks for testing. Can you not just also add `127.0.0.1` to the SAN list?

**Comment by Safihre** [comment]
@thezoggy IPv6 is causing problems:
```
Traceback (most recent call last):
  File "C:\Users\<USERNAME>\Documents\GitHub\sabnzbd\sabnzbd\misc.py", line 1254, in create_https_certificates
    cert = generate_local_cert(private_key, days_valid=3560, output_file=ssl_cert, LN=u'SABnzbd', ON=u'SABnzbd', CN=u'localhost')
  File "C:\Users\<USERNAME>\Documents\GitHub\sabnzbd\sabnzbd\utils\certgen.py", line 89, in generate_local_cert
    ).sign(private_key, hashes.SHA256(), default_backend())
  File "C:\Python27\Lib\site-packages\cryptography\x509\base.py", line 564, in sign
    return backend.create_x509_certificate(self, private_key, algorithm)
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\multibackend.py", line 404, in create_x509_certificate
    return b.create_x509_certificate(builder, private_key, algorithm)
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\backend.py", line 883, in create_x509_certificate
    gc=True
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\backend.py", line 1009, in _create_x509_extensions
    handlers, extension
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\backend.py", line 1040, in _create_x509_extension
    ext_struct = encode(self, extension.value)
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\encode_asn1.py", line 359, in _encode_alt_name
    general_names = _encode_general_names(backend, san)
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\encode_asn1.py", line 351, in _encode_general_names
    gn = _encode_general_name(backend, name)
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\encode_asn1.py", line 387, in _encode_general_name
    value = _idna_encode(name.value)
  File "C:\Python27\Lib\site-packages\cryptography\hazmat\backends\openssl\encode_asn1.py", line 376, in _idna_encode
    return idna.encode(value)
  File "C:\Python27\Lib\site-packages\idna\core.py", line 355, in encode
    result.append(alabel(label))
  File "C:\Python27\Lib\site-packages\idna\core.py", line 276, in alabel
    check_label(label)
  File "C:\Python27\Lib\site-packages\idna\core.py", line 253, in check_label
    raise InvalidCodepoint('Codepoint {0} at position {1} of {2} not allowed'.format(_unot(cp_value), pos+1, repr(label)))
InvalidCodepoint: Codepoint U+003A at position 5 of u'2001:0:34a9:ece7:82e:1f23:abe7:dd15' not allowed
```

**Comment by thezoggy** [comment]
@Safihre yeah.. its because crypto wants unicode argument for dns name so it can decode.. but doing so for v6 turns the colon  into unicode which then gets disallowed.. blah I'll just use the not so well known IPAddress type but then it involves having to convert the string to ipaddress type...  one sec. ill add 127.0.0.1 as well for ya

**Comment by thezoggy** [comment]
so if I use x509.DNSName the apps are happy with 127.0.0.1/localhost/localip.. if I use x509.IPAddress for 127.0.0.1 or localip.. it doesnt validate...

**Comment by thezoggy** [comment]
yeah, so you must pass a unicode string to crypto for dnsname.. but that name cant have a colon.. thus we wont be able to do the SAN for the v6 address.. we could add it under ipaddress type but as mentioned before.. its rarely used and in my case radarr/sonarr arent validating the certs off them (only dns name)..  so honestly those using v6 will have to just create their own cert since I cant find a way to sidestep crypto's requirement without monkey patching and I really dont want to go down this road

**Comment by thezoggy** [comment]
alright pushed update, let me know when you want me to rebase+squash and force push. i just cant get v6 addresses to work in the dnsname.. due to the requirement of unicode string being passed but then it failing due to colons in it...

**Comment by Safihre** [comment]
Oke looks good. I was only wondering why you leave out the `CN` completly? Why not put `127.0.0.1` or `localhost` there?

**Comment by thezoggy** [comment]
if we include CN lots of app will check that and only that.. not including it forces them to use SAN..  CN is legacy way to validate.. 

per RFC 6125, published in '2011 the validator must check SAN first, and if SAN exists, then CN should not be checked. Now not all apps follow this, and some will use the CN if its there ignoring the SAN.. but if CN isnt there.. it forces it to use SAN. So this is to ensure things are checked correctly. Also I noticed that when I import the cert into the trusted root store that if I provide a CN it shows up as that.. if I dont provide CN it uses the issuer (sabnzbd) which looks much nicer.

**Comment by Safihre** [comment]
Shouldn't the issuer have a `CN`? Not that it really matters of course..

**Comment by thezoggy** [comment]
Pre 2011 yes, with our use (multiple domains/self-signed) CN would not be good..

**Comment by Safihre** [comment]
No I mean `CN` for issuer, not subject.

**Comment by thezoggy** [comment]
the CN is for the issuer.. which we purposely dont want to use.. because we want to use the SubjectAltName

**Comment by thezoggy** [comment]
btw, the certgen code it makes the subject and the issuer the same which is the norm for self-signed certs... we could put a CN for the subject and not the issuer but honestly if we were really trying to create a legit cert we also need to put state/country/email/etc.. but all that stuff is moot since the CA issuer for self-sign certs can never be valid due to their nature.. thus it doesnt matter. the main goal was just to make other local apps that validate the cert to sab happier.. which this now does.

**Pull request merged**